dom.go

```diff
@@ -252,19 +268,59 @@ func (h *HTML) reconcile(prev *HTML) []Mounter {
 		h.node.Set("innerHTML", h.innerHTML)
 	}
 
-	return h.reconcileChildren(prev, nil)
+	return h.reconcileChildren(prev)
 }
 
 // reconcileChildren reconciles children of the current HTML against a previous
 // render's DOM nodes.
-func (h *HTML) reconcileChildren(prev *HTML, insertBefore *HTML) (pendingMounts []Mounter) {
-	// TODO better list element reuse
+func (h *HTML) reconcileChildren(prev *HTML) (pendingMounts []Mounter) {
+	var keyed bool
 	for i, nextChild := range h.children {
-		// TODO(pdf): Add tests for node equality
-		new := h.node != prev.node
-		if i >= len(prev.children) || new {
-			if nextChildList, ok := nextChild.(List); ok {
-				nextChildList.reconcile(h.node, insertBefore, nil)
+		// Massage concrete type if necessary.
+		switch v := nextChild.(type) {
+		case *HTML:
+			if v == nil {
+				nextChild = nil
+				h.children[i] = nextChild
+			}
+		case List:
+			nextChild = childList{html: &HTML{children: v}}
+			h.children[i] = nextChild
```

Comment:
`h.children`'s data type is `[]ComponentOrHTML` which does not include `childList`. It feels wrong to ignore this.

Reply:
ComponentOrHTML is exported, but this is an unexported type, what would you like to do?

Reply:
Could we change it to an `interface{}` type without requiring extension copying / conversion all over the place? (I am not sure).

Otherwise, maybe we just need a doc comment on the definition of `children` now:

```
// important: children is actually ComponentOrHTML OR childList here, from an internal perspective.
```